crates/sui-graphql-rpc/src/metrics.rs

```diff
@@ -1,15 +1,209 @@
+                        if val == code::INTERNAL_SERVER_ERROR {
+                            self.request_metrics
+                                .num_internal_errors_by_path
+                                .with_label_values(&[&self.get_query_path(err.path)])
+                                .inc();
+                        } else if val == code::BAD_USER_INPUT {
+                            self.request_metrics
+                                .num_bad_input_errors_by_path
+                                .with_label_values(&[&self.get_query_path(err.path)])
+                                .inc();
+                        } else {
+                            self.request_metrics
+                                .num_query_errors_by_path
+                                .with_label_values(&[&self.get_query_path(err.path)])
+                                .inc();
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Use this function to increment the number of request error.
+    /// If you need to increase error rate due to bad input, use both
+    /// this function and `inc_error_bad_input`
+    pub(crate) fn inc_num_query_success(&self) {
+        self.request_metrics
+            .num_query_success
+            .with_label_values(&["path"])
+            .inc();
+    }
+
+    /// When an error occurs, GraphQL returns a vector of PathSegments,
+    /// that we can use to construct a simplified path to the actual error.
+    pub(crate) fn get_query_path(&self, query: Vec<PathSegment>) -> String {
+        let mut path = String::new();
+        for (idx, s) in query.iter().enumerate() {
+            if idx > 0 {
+                path.push('.');
+            }
+            match s {
+                PathSegment::Index(idx) => {
+                    let _ = write!(&mut path, "{}", idx);
+                }
+                PathSegment::Field(name) => {
+                    let _ = write!(&mut path, "{}", name);
+                }
+            }
+        }
+        path
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct DBMetrics {
+    pub db_num_fetches: IntCounterVec,
+    pub db_fetch_latency: HistogramVec,
+    // TODO make this work
+    pub _db_query_cost: Histogram,
+    // TODO determine if we want this metric, and implement it
+    pub _db_fetch_batch_size: HistogramVec,
+}
+
+const LATENCY_SEC_BUCKETS: &[f64] = &[
+    0.0001, 0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.5, 1., 2.5, 5., 10., 20., 30., 60., 90.,
+];
+
+const BATCH_SIZE_BUCKETS: &[f64] = &[
+    1., 2., 4., 8., 12., 16., 24., 32., 48., 64., 96., 128., 256., 512., 1024.,
+];
+
+impl DBMetrics {
+    pub(crate) fn new(registry: &Registry) -> Self {
+        Self {
+            db_num_fetches: register_int_counter_vec_with_registry!(
+                "db_fetch",
+                "The total number of DB requests grouped by success or error",
```

Comment:
Imo success count is not useful, you usually end up ignoring it

Reply:
I am still learning the best practices here, so happy to hear other folks' experience. We can either do `db_fetch` as number of total DB requests, and then `db_fetch_error` to understand the error rate, or we can have them as `success` and `error` which is the total num of db fetches, from which we can also compute the error. I thought one field instead of two might be easier.